Add installation instructions for M1 Mac. (#1724)
NNPDF/nnpdf: doc/sphinx/source/get-started/installation.rst

.. _anconda documentation: https://docs.conda.io/projects/conda-build/en/latest/resources/compiler-tools.html#macos-sdk

.. _M1:

Installation from source on M1 Macs
-----------------------------------

Installation on M1 Macs is not directly supported, so everything needs to be
done manually. The following steps are required:

1. Clone repos

  .. code::

      mkdir nnpdfgit
      cd nnpdfgit
      git clone git@github.com:NNPDF/nnpdf.git
      git clone git@github.com:NNPDF/binary-bootstrap.git
      git clone https://github.com/scarrazza/apfel.git

> 2. Execute binary bootstrap to set channels in .condarc

Reviewer: double quotes probably

> 2. Execute binary bootstrap to set channels in .condarc

Author: done

> 2. Execute binary bootstrap to set channels in .condarc
>    .. code::

Reviewer: ```suggestion
2. Execute binary bootstrap to set channels in .condarc

   .. code::
```

      ./binary-bootstrap/bootstrap.sh

3. Setup conda environment using python 3.9

   .. code::

      conda create -n nnpdf-dev python=3.9
      conda activate nnpdf-dev

4. Install compiler

   .. code::

      conda install clangxx_osx-arm64

> 5. LHAPDF

Author: No, we did this either in step 5 or 6, I forget now, and I found that it could be done instead simply with `autoreconf -f -i`

> 5. LHAPDF

Reviewer 2: Ah ok, makes sense.

   Download version 6.4.0 and decompress

    .. code::

       wget -O LHAPDF-6.4.0.tar.gz https://lhapdf.hepforge.org/downloads/?f=LHAPDF-6.4.0.tar.gz
       tar -xzvf LHAPDF-6.4.0.tar.gz
       rm LHAPDF-6.4.0.tar.gz
       cd LHAPDF-6.4.0

   Regenerate the configuration files, configure the build with the specified options, and compile and install the software.
   You may need to `brew install automake` first:

    .. code::

      autoreconf -f -i
      ./configure --prefix=$CONDA_PREFIX --disable-python

>       make -j8

Reviewer: ```suggestion
      make -j
```

      make install

   Install python wrapper

   .. code::

      cd wrappers/python
      pip install -e .

>    Test
>    .. code::

Reviewer: ```suggestion
   Test

   .. code::
```

      lhapdf install CT18NNLO

>       python
> 
>    Just check if `import lhapdf` works

Reviewer: ```suggestion
      python -c "import lhapdf"
```

6. Apfel

   Not sure if this is necessary, but to make sure the environment is active,

   .. code::

      echo 'export PATH=${CONDA_PREFIX}/bin:$PATH' >> "${CONDA_PREFIX}/etc/conda/activate.d/env_vars.sh"
      conda deactivate
      conda activate nnpdf-dev

   install these:

   .. code::
      conda install pkg-config swig cmake

>    Then build it
>    .. code::

Reviewer: ```suggestion
   Then build it

   .. code::
```

      cd ../../../apfel
      autoreconf -f -i
      PYTHON=$(which python) ./configure --prefix=$CONDA_PREFIX 
      make clean
      make -j8
      make install

7. validphys

   First install reportengine and validobj, then validphys itself:

   .. code::

      pip install reportengine validobj
      cd ../nnpdf/validphys2
      pip install -e .

8. nnpdf

   Install other dependencies

   .. code::

      conda install libarchive sqlite gsl yaml-cpp

   Run cmake in nnpdf/build directory:

   .. code::

      cd ..
      mkdir build
      cd build
      cmake .. -DCMAKE_INSTALL_PREFIX=$CONDA_PREFIX

>    Edit the file `nnpdfgit/nnpdf/CMakeLists.txt` :

Reviewer: ```suggestion
   Edit the file ``nnpdfgit/nnpdf/CMakeLists.txt`` :
```

    - on line 8 change the option to true, so it says:

      .. code::

         SET(CMAKE_BUILD_WITH_INSTALL_RPATH TRUE)

>     - comment out line 58 (`set(LIBNNPDF_HAVE_SSE "#define SSE_CONV")`)

Reviewer: ```suggestion
    - comment out line 58 (:code:`set(LIBNNPDF_HAVE_SSE "#define SSE_CONV")`)
```

    - line 104 should read:

      .. code::

         set(DEFAULT_CXX_OPTIONS "-Wall -Wextra -fvisibility-inlines-hidden -fmessage-length=0 -ftree-vectorize -fPIC -fstack-protector-strong -O2 -pipe")

>        (so delete "-march=nocona -mtune=haswell").

Reviewer: ```suggestion
      (so delete ``-march=nocona -mtune=haswell``).
```

   Then make:

   .. code::

      make -j8
      make install

>    Install remaining packages
>    .. code::

Reviewer: ```suggestion
   Install remaining packages

   .. code::
```

      pip install seaborn prompt_toolkit scipy psutil hyperopt

> 9. Install tensorflow
>    Not specifying versions will install at the time of writing macos 2.12.0 and metal 0.8.0, which both work.

Reviewer: ```suggestion
9. Install tensorflow

   Not specifying versions will install at the time of writing macos 2.12.0 and metal 0.8.0, which both work.
```

   They only give warnings on the optimizers, that the legacy versions are faster.
   If you want an older version, macos 2.9.2 and metal 0.5.0 are also tested to work.

   .. code::

      conda install -c apple tensorflow-deps
      pip install tensorflow-macos==2.9.2
      pip install tensorflow-metal==0.5.0

10. Test

   .. code::

      cd nnpdf/n3fit/runcards/examples
      vp-setupfit Basic_runcard.yml
      n3fit Basic_runcard.yml 1
      evolven3fit Basic_runcard 1

>    You can experiment with setting
> 
>    .. code::
> 
>       tf.config.set_visible_devices([], 'GPU')
> 
>    to disable the GPU for tensorflow. In this small example at least, this makes it faster.

Reviewer 2: ```suggestion
   With these settings tensorflow will run by default on GPU which makes
   the fit run very slow. To disable the GPU, type the following command:
   
   .. code::

      export CUDA_VISIBLE_DEVICES=0
      
   or insert the following line in ... [followed by Roy's comment]

   .. code::

      tf.config.set_visible_devices([], 'GPU')
```

>    You can add this line for example in the `set_initial_state` function in `n3fit/src/n3fit/backends/keras_backend/internal_state.py`.
>    And to use legacy optimizers, you only need to change one line in `n3fit/src/n3fit/backends/keras_backend/MetaModel.py`:

Reviewer: ```suggestion
   You can add this line for example in the ``set_initial_state`` function in ``n3fit/src/n3fit/backends/keras_backend/internal_state.py``.
   And to use legacy optimizers, you only need to change one line in ``n3fit/src/n3fit/backends/keras_backend/MetaModel.py``:
```

.. _anconda documentation: https://docs.conda.io/projects/conda-build/en/latest/resources/compiler-tools.html#macos-sdk

.. _M1:

Installation from source on M1/M2 Macs

> -----------------------------------

Reviewer: ```suggestion
--------------------------------------
```

2. Execute binary bootstrap to set the channels in ``.condarc``

   .. code::

      ./binary-bootstrap/bootstrap.sh

3. Setup conda environment using python 3.9

   .. code::

      conda create -n nnpdf-dev python=3.9
      conda activate nnpdf-dev

4. Install ARM compiler

   .. code::

      conda install clangxx_osx-arm64

5. LHAPDF

   Download version 6.4.0 and decompress

    .. code::

       wget -O LHAPDF-6.4.0.tar.gz https://lhapdf.hepforge.org/downloads/?f=LHAPDF-6.4.0.tar.gz
       tar -xzvf LHAPDF-6.4.0.tar.gz
       rm LHAPDF-6.4.0.tar.gz
       cd LHAPDF-6.4.0

   Regenerate the configuration files, configure the build with python disabled, compile and 
   install. You may need to `brew install automake` first:

    .. code::

      autoreconf -f -i
      ./configure --prefix=$CONDA_PREFIX --disable-python
      make -j
      make install

   Install the python wrapper

   .. code::

      cd wrappers/python
      pip install -e .

   Test

   .. code::

      lhapdf install CT18NNLO
      python -c "import lhapdf"

6. Apfel

   First, we need to install some dependencies:

   .. code::
      conda install pkg-config swig cmake

   Then build it

>    .. code::
>       cd ../../../apfel

Reviewer: ```suggestion
   .. code::

      cd ../../../apfel
```

      (so delete ``-march=nocona -mtune=haswell``).

   Then make:

   .. code::

      make -j
      make install

   Install remaining packages

   .. code::

      pip install seaborn prompt_toolkit scipy psutil hyperopt

9. Install tensorflow

   Not specifying versions will install at the time of writing macos 2.12.0 and metal 0.8.0, which both work.
   They only give warnings on the optimizers, that the legacy versions are faster.
   If you want an older version, macos 2.9.2 and metal 0.5.0 are also tested to work.

   .. code::

      conda install -c apple tensorflow-deps
      pip install tensorflow-macos==2.9.2
      pip install tensorflow-metal==0.5.0

10. Test

   .. code::

      cd nnpdf/n3fit/runcards/examples
      vp-setupfit Basic_runcard.yml
      n3fit Basic_runcard.yml 1
      evolven3fit Basic_runcard 1

   With these settings tensorflow will run by default on GPU which makes
   the fit run very slow. To disable the GPU, type the following command:
   
   .. code::

      export CUDA_VISIBLE_DEVICES=0
      

>    or insert the following line in ... [followed by Roy's comment]

Reviewer: Not sure what my comment was, but I guess it shouldn't be left like this.

Reviewer 2: Yes, it definitely shouldn't be left like this. It should read:

```rst
or insert the following line in in the ``set_initial_state`` function in ``n3fit/src/n3fit/backends/keras_backend/internal_state.py``:

.. code::

      tf.config.set_visible_devices([], 'GPU')
```

And then followed by:
```rst
And to use legacy optimizers, ...
```

> .. code::

Reviewer: ```suggestion
    .. code::
```

>       tf.config.set_visible_devices([], 'GPU')
>    You can add this line for example in the ``set_initial_state`` function in ``n3fit/src/n3fit/backends/keras_backend/internal_state.py``.

Reviewer: ```suggestion
      tf.config.set_visible_devices([], 'GPU')

   You can add this line for example in the ``set_initial_state`` function in ``n3fit/src/n3fit/backends/keras_backend/internal_state.py``.
```

Author: Sorry should have caught that, fixed
